src/modules/schnorrsig_halfagg/tests_impl.h

```diff
@@ -0,0 +1,347 @@
+        unsigned char sk[32];
+        secp256k1_keypair keypair;
+        secp256k1_testrand256(sk);
+        secp256k1_testrand256(&msgs32[i*32]);
+
+        CHECK(secp256k1_keypair_create(CTX, &keypair, sk));
+        CHECK(secp256k1_keypair_xonly_pub(CTX, &pubkeys[i], NULL, &keypair));
+        CHECK(secp256k1_schnorrsig_sign(CTX, &sigs64[i*64], &msgs32[i*32], &keypair, NULL));
+    }
+}
+
+/* In this test we create a bunch of Schnorr signatures,
+ * aggregate some of them in one shot, and then
+ * aggregate the others incrementally to the already aggregated ones.
+ * The aggregate signature should verify after both steps. */
+void test_schnorrsig_aggregate(void) {
+    secp256k1_xonly_pubkey pubkeys[N_MAX];
+    unsigned char msgs32[N_MAX*32];
+    unsigned char sigs64[N_MAX*64];
+    unsigned char aggsig[32*(N_MAX + 1) + 17];
+    size_t aggsig_len = sizeof(aggsig);
+
+    size_t n = secp256k1_testrand_int(N_MAX + 1);
+    size_t n_initial = secp256k1_testrand_int(n + 1);
+    size_t n_new = n - n_initial;
+    test_schnorrsig_aggregate_input_helper(pubkeys, msgs32, sigs64, n);
+
+    /* Aggregate the first n_initial of them */
+    CHECK(secp256k1_schnorrsig_aggregate(CTX, aggsig, &aggsig_len, pubkeys, msgs32, sigs64, n_initial));
+    /* Make sure that the aggregate signature verifies */
+    CHECK(aggsig_len == 32*(n_initial + 1));
+    CHECK(secp256k1_schnorrsig_aggverify(CTX, pubkeys, msgs32, n_initial, aggsig, aggsig_len));
+    /* Aggregate the remaining n_new many signatures to the already existing ones */
+    aggsig_len = sizeof(aggsig);
+    secp256k1_schnorrsig_inc_aggregate(CTX, aggsig, &aggsig_len, pubkeys, msgs32, &sigs64[n_initial*64], n_initial, n_new);
+    /* Make sure that the aggregate signature verifies */
+    CHECK(aggsig_len == 32*(n + 1));
+    CHECK(secp256k1_schnorrsig_aggverify(CTX, pubkeys, msgs32, n, aggsig, aggsig_len));
+
+    /* Check that a direct aggregation of the n sigs yields an identical aggsig */
+    {
+        unsigned char aggsig2[sizeof(aggsig)];
+        size_t aggsig_len2 = sizeof(aggsig2);
+        CHECK(secp256k1_schnorrsig_aggregate(CTX, aggsig2, &aggsig_len2, pubkeys, msgs32, sigs64, n));
+        CHECK(aggsig_len == aggsig_len2);
+        CHECK(secp256k1_memcmp_var(aggsig, aggsig2, aggsig_len) == 0);
+    }
+}
+
+/* This tests the verification test vectors from
+ * https://github.com/BlockstreamResearch/cross-input-aggregation/blob/master/hacspec-halfagg/tests/tests.rs#L78 . */
+void test_schnorrsig_aggverify_spec_vectors(void) {
+    /* Test vector 0 */
+    {
+        size_t n = 0;
+        const unsigned char pubkeys_ser[32] = {
+            0x1b, 0x84, 0xc5, 0x56, 0x7b, 0x12, 0x64, 0x40,
+            0x99, 0x5d, 0x3e, 0xd5, 0xaa, 0xba, 0x05, 0x65,
+            0xd7, 0x1e, 0x18, 0x34, 0x60, 0x48, 0x19, 0xff,
+            0x9c, 0x17, 0xf5, 0xe9, 0xd5, 0xdd, 0x07, 0x8f
+        };
+        secp256k1_xonly_pubkey pubkeys[1]; /* C doesn't allow zero-size arrays. */
```

Comment:
Note: This indicates that the API is a bit inconvenient for empty aggregates (i.e., n=0). 

In the schnorrsig module, we allow the caller to pass NULL for an empty message:
https://github.com/bitcoin-core/secp256k1/blob/a9db9f2d75ac9df0312dec0c329266969bcc2946/include/secp256k1_schnorrsig.h#L154

We could probably do the same here, if only for consistency. I am not sure if any caller will need this, but from the point of view of the implementation it's really just the promise not to deference the pointer (which we already keep!),  and a modified ARG_CHECK.

But let's first see if we would like to allow n=0 at all https://github.com/BlockstreamResearch/cross-input-aggregation/issues/11#issuecomment-1877084001

Reply:
> But let's first see if we would like to allow n=0 at all [BlockstreamResearch/cross-input-aggregation#11 (comment)](https://github.com/BlockstreamResearch/cross-input-aggregation/issues/11#issuecomment-1877084001)

I think we want to allow n=0...  @jonasnick Do you think we should document that NULL is allowed here for n=0? I tend to say yes, just because it doesn't seem to hurt.

Reply:
Yeah, we can change the ARG_CHECKs in aggverify to
```
ARG_CHECK(pubkeys != NULL || n == 0);
ARG_CHECK(msgs32 != NULL || n == 0);
```
and change the function doc accordingly.

Reply:
Shouldn't it be:
```
ARG_CHECK(pubkeys != NULL || (n == 0 && pubkeys == NULL));
ARG_CHECK(msgs32  != NULL || (n == 0 && msgs32  == NULL));
ARG_CHECK(aggsig  != NULL || (n == 0 && aggsig  == NULL));
```
Although in the Schnorrsig module it seems like the public key and signature must never be NULL.
I don't know if we want to allow that n = 0 but pk / msg / aggsig is not NULL.

Reply:
I think they're logically equivalent:
```
pubkeys != NULL || (n == 0 && pubkeys == NULL)
= (pubkeys != NULL || n == 0) && (pubkeys != NULL || pubkeys == NULL)
= pubkeys != NULL || n == 0
```

I don't think we need to allow `aggsig == NULL`. For `n = 0`, aggsig is defined to be 32 bytes of 0.

Reply:
Oh, sure. 
But if `msgs32 == NULL`, do we want to require that `n > 0`?

Reply:
> But if `msgs32 == NULL`, do we want to require that `n > 0`?

I don't fully understand. You mean `msgs32 != NULL`?

Anyway, what @jonasnick suggested seems sane to me:
```
ARG_CHECK(pubkeys != NULL || n == 0);
ARG_CHECK(msgs32 != NULL || n == 0);
```
One can think of the inputs as (pubkey, message) pairs. (And this is how you'd implement it in other languages.) So it makes sense to treat `pubkeys` and `msgs32` equally. But the signature should always be present.

And note that is different from the Schnorr module. There, `msg32 == NULL` merely indicates a message of `""` (byte array of length 0), but the message is still present. (And we can't do this here, this would require a different interface where `msgs32` is an `unsigned char **`. Such an interface may save the caller some memcpys, but let's not care about this now.)

Reply:
> I don't fully understand. You mean msgs32 != NULL?

yes, that's what I meant, sorry.
But I will implement @jonasnick's suggestion then.